### PR TITLE
8283346: Optimize observable ArrayList creation in FXCollections

### PR DESCRIPTION
**Pull request by Maran23**
This simple PR optimizes the observable `ArrayList` creation by using the ArrayList constructor/array size so that the underlying array will be initialized at the correct size which will speed up the creation as the array does not need to grow as a result of the `addAll` call.

I also added tests which will succeed before and after to verify that nothing got broken by this change.
Also I made a benchmark test. Results:

| Benchmark | Mode| Cnt | Score | Error | Units
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| ListBenchmark OLD  | thrpt | 25 | 722,842 | ± 26,93 | ops/s
| ListBenchmark NEW | thrpt  | 25 | 29262,274 | ± 2088,712 | ops/s

Edit: I also made a synthetic benchmark by measuring the same code below 100 times with `System.nanoTime`.
ListBenchmark OLD (avg): 21-23ms
ListBenchmark NEW (avg): 2 ms

<details><summary>Benchmark code</summary>

```
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.TearDown;

import java.util.ArrayList;
import java.util.List;

@State(Scope.Benchmark)
public class ListBenchmark {

    List<String> strings;

    @Setup
    public void setup() {
        strings = new ArrayList<>();
        for(int i = 0; i< 100000;i++) {
            strings.add("abc: " + i);
        }
    }

    @TearDown
    public void tearDown() {
        strings = null;
    }

    @Benchmark
    public ObservableList<String> init() {
        return FXCollections.observableArrayList(strings);
    }
}
```

</details>

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8283346](https://bugs.openjdk.org/browse/JDK-8283346): Optimize observable ArrayList creation in FXCollections

### Reviewers
 * [Michael Strauß](https://openjdk.org/census#mstrauss) (@mstr2 - Author)
 * [Nir Lisker](https://openjdk.org/census#nlisker) (@nlisker - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/758/head:pull/758` \
`$ git checkout pull/758`

Update a local copy of the PR: \
`$ git checkout pull/758` \
`$ git pull https://git.openjdk.org/jfx pull/758/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 758`

View PR using the GUI difftool: \
`$ git pr show -t 758`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/758.diff">https://git.openjdk.org/jfx/pull/758.diff</a>

</details>

### CODE REVIEW

**Comment by yososs** [comment]
This PR appears to have two internal behavior changes.

A. Initial collection size (PR owner's description)
B. By adding the item directly to the backingList
Avoiding beginChange () --doAdd () --endChange ()

``` java
     @Override
     protected void doAdd (int index, E element) {
         if (elementObserver! = null)
             elementObserver.attachListener (element);
         backingList.add (index, element);
     }
```

Looking at the following code, it seems that there is no problem because attachListener () is called in the constructor.
We will need to make sure that the current unit tests cover B's changes.

ObservableListWrapper.java
``` java
    public ObservableListWrapper(List<E> list) {
        backingList = list;
        elementObserver = null;
    }

    public ObservableListWrapper(List<E> list, Callback<E, Observable[]> extractor) {
        backingList = list;
        this.elementObserver = new ElementObserver(extractor, new Callback<E, InvalidationListener>() {

            @Override
            public InvalidationListener call(final E e) {
                return new InvalidationListener() {

                    @Override
                    public void invalidated(Observable observable) {
                        beginChange();
                        int i = 0;
                        final int size = size();
                        for (; i < size; ++i) {
                            if (get(i) == e) {
                                nextUpdate(i);
                            }
                        }
                        endChange();
                    }
                };
            }
        }, this);
        final int sz = backingList.size();
        for (int i = 0; i < sz; ++i) {
            elementObserver.attachListener(backingList.get(i));
        }
    }
```

**Comment by yososs** [comment]
This PR appears to have two internal behavior changes.

A. Initial collection size (PR owner's description)
B. By adding the item directly to the backingList
Avoiding beginChange () --doAdd () --endChange ()

``` java
     @Override
     protected void doAdd (int index, E element) {
         if (elementObserver! = null)
             elementObserver.attachListener (element);
         backingList.add (index, element);
     }
```

Looking at the following code, it seems that there is no problem because attachListener () is called in the constructor.
We will need to make sure that the current unit tests cover B's changes.

ObservableListWrapper.java
``` java
    public ObservableListWrapper(List<E> list) {
        backingList = list;
        elementObserver = null;
    }

    public ObservableListWrapper(List<E> list, Callback<E, Observable[]> extractor) {
        backingList = list;
        this.elementObserver = new ElementObserver(extractor, new Callback<E, InvalidationListener>() {

            @Override
            public InvalidationListener call(final E e) {
                return new InvalidationListener() {

                    @Override
                    public void invalidated(Observable observable) {
                        beginChange();
                        int i = 0;
                        final int size = size();
                        for (; i < size; ++i) {
                            if (get(i) == e) {
                                nextUpdate(i);
                            }
                        }
                        endChange();
                    }
                };
            }
        }, this);
        final int sz = backingList.size();
        for (int i = 0; i < sz; ++i) {
            elementObserver.attachListener(backingList.get(i));
        }
    }
```

**Comment by mstr2** [comment]
The following pieces of code should be identical:
```java
var list = FXCollections.observableArrayList();
list.addAll(source);
```
```java
var list = FXCollections.observableList(new ArrayList<>(source));
```
Any observable difference in behavior would be unspecified. When it comes to `modCount`, that's an internal field of `AbstractList`.
`FXCollections.observableArrayList()` and `FXCollections.observableList(...)` are not the right place to test this.

**Comment by Maran23** [comment]
> * I don't know what the modCount property is used for, but is it okay if the modified code has different values?

This is fine as the mod count is only there to detect concurrent modifications. 
I recommend to read the `modCount` javadoc, its is very well written.
And as @mstr2 also wrote, it is also an internal field (for the above mentioned purpose).

File: modules/javafx.base/src/main/java/javafx/collections/FXCollections.java [hunk 1]
```diff
@@ -347,7 +347,7 @@ public static <E> ObservableList<E> observableArrayList(Callback<E, Observable[]
      * @see #observableArrayList()
      */
     public static <E> ObservableList<E> observableArrayList(E... items) {
-        ObservableList<E> list = observableArrayList();
+        ObservableList<E> list = observableList(new ArrayList<>(items.length));
         list.addAll(items);
```
[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.base/src/main/java/javafx/collections/FXCollections.java]: done

File: modules/javafx.base/src/main/java/javafx/collections/FXCollections.java [hunk 2]
```diff
@@ -347,9 +347,9 @@ public static <E> ObservableList<E> observableArrayList(Callback<E, Observable[]
      * @see #observableArrayList()
      */
     public static <E> ObservableList<E> observableArrayList(E... items) {
-        ObservableList<E> list = observableArrayList();
-        list.addAll(items);
-        return list;
+        ArrayList<E> backingList = new ArrayList<>(items.length);
+        backingList.addAll(Arrays.asList(items));
```
Comment (nlisker) [inline modules/javafx.base/src/main/java/javafx/collections/FXCollections.java:351]: Did you try using `List.of` instead?
```java
var backingList = new ArrayList<>(List.of(items));
```

Comment (Maran23) [reply modules/javafx.base/src/main/java/javafx/collections/FXCollections.java:351]: No, but `List.of` does not allow `null` values, therefore this would change the existing behaviour as `null` is allowed in an observable arraylist.

**Review by mstr2 — APPROVED** [review]

File: modules/javafx.base/src/test/java/test/javafx/collections/FXCollectionsTest.java [hunk 3]
```diff
@@ -42,6 +42,47 @@
 
 public class FXCollectionsTest {
 
+    @Test
+    public void testCreateObservableArrayListFromArray() {
+        ObservableList<String> observableList = FXCollections.observableArrayList("1", "2", "3");
```
Comment (nlisker) [inline modules/javafx.base/src/test/java/test/javafx/collections/FXCollectionsTest.java:47]: I would add a `null` element (like `"1", "2", null`) to test that the method accepts `null`s, especially after I fell for it myself.

File: modules/javafx.base/src/test/java/test/javafx/collections/FXCollectionsTest.java [hunk 4]
```diff
@@ -42,6 +42,47 @@
...
+
+        assertEquals(3, observableList.size());
+        assertTrue(observableList.contains("1"));
+        assertTrue(observableList.contains("2"));
+        assertTrue(observableList.contains("3"));
+    }
+
+    @Test
+    public void testCreateObservableArrayListFromCollection() {
+        List<String> list = List.of("1", "2", "3");
```
Comment (nlisker) [inline modules/javafx.base/src/test/java/test/javafx/collections/FXCollectionsTest.java:57]: Same here with `null`, only now `Arrays.asList` will be needed instead.

File: modules/javafx.base/src/main/java/javafx/collections/FXCollections.java (hunk 2 above)
Comment (nlisker) [inline modules/javafx.base/src/main/java/javafx/collections/FXCollections.java:351]: Unless I'm missing something again, this can be written as
`ArrayList<E> backingList = new ArrayList<>(Arrays.asList(items));`

I created benchmarks based on yours, and there is no significant difference between these two (their scores are within the error ranges of each other).

I don't mind leaving it as is if you think it's clearer, I personally prefer the 1 line variant.

**Review by nlisker — COMMENTED** [review]
Looks good. I left a few small comments.

I ran some benchmarks and I can reproduce the performance improvement both in the array (varargs) and the collection variants of the method.

File: modules/javafx.base/src/main/java/javafx/collections/FXCollections.java (hunk 2 above)
Comment (Maran23) [reply modules/javafx.base/src/main/java/javafx/collections/FXCollections.java:351]: Thanks. I totally missed that, I prefer the 1 line variant as well.

File: modules/javafx.base/src/test/java/test/javafx/collections/FXCollectionsTest.java (hunk 3 above)
Comment (Maran23) [reply modules/javafx.base/src/test/java/test/javafx/collections/FXCollectionsTest.java:47]: Good idea, will do.

**Comment by Maran23** [comment]
> Looks good. I left a few small comments.
> 
> I ran some benchmarks and I can reproduce the performance improvement both in the array (varargs) and the collection variants of the method.

Thanks for the review and verifying the improvement.
I implemented your changes.

**Review by nlisker — APPROVED** [review]

**Review by mstr2 — APPROVED** [review]

**Review by kevinrushforth — APPROVED** [review]
